src/spark_rapids_ml/knn.py

```diff
@@ -493,3 +521,100 @@ def _get_cuml_transform_func(
+
+        numNearestNeighbors: int
+            the number of exact nearest neighbors to be returned for each query.
+
+        distCol: str
+            the name of the distance column
+
+        Returns
+        -------
+        knnjoin_df: pyspark.sql.DataFrame
+            the result dataframe that has three columns (datasetA, datasetB, distCol).
+            datasetA column is of struct type that concatenates all the columns of input dataframe datasetA.
+            Similarly, datasetB column is of struct type that concatenates all the columns of input dataframe datasetB.
+            distCol is the distance column. A row in knnjoin_df is in the format (v1, v2, dist(v1, v2)),
+            where item_vector v1 is one of the k nearest neighbors of query_vector v2 and their distance is dist(v1, v2).
+        """
+
+        # save relevant self variables
+        estimator_item_df_withid = self._item_df_withid
+        estimator_processed_item_df = self._processed_item_df
+        estimator_n_neighbors = self.cuml_params["n_neighbors"]
+
+        # modify relevant self variables
+        id_col_name = self.getIdCol()
+
+        self._item_df_withid = self.ensureIdCol(datasetA)
+
+        self._processed_item_df = self._item_df_withid.withColumn(
+            alias.label, lit(self._label_isdata)
+        )
+        self.cuml_params["n_neighbors"] = numNearestNeighbors
+
+        # call kneighbors then prepare return results
+        (item_df_withid, query_df_withid, knn_df) = self.kneighbors(datasetB)
+
+        from pyspark.sql.functions import arrays_zip, col, explode, struct
+
+        knn_pair_df = knn_df.select(
+            f"query_{id_col_name}",
+            explode(arrays_zip("indices", "distances")).alias("zipped"),
+        ).select(
+            f"query_{id_col_name}",
+            col("zipped.indices").alias(f"item_{id_col_name}"),
+            col("zipped.distances").alias(distCol),
+        )
+
+        datasetA_res = item_df_withid.select(struct("*").alias("datasetA"))
+        datasetB_res = query_df_withid.select(struct("*").alias("datasetB"))
+
+        knnjoin_df = datasetA_res.join(
+            knn_pair_df,
+            datasetA_res[f"datasetA.{id_col_name}"]
+            == knn_pair_df[f"item_{id_col_name}"],
+        )
+        knnjoin_df = knnjoin_df.join(
+            datasetB_res,
+            knnjoin_df[f"query_{id_col_name}"]
+            == datasetB_res[f"datasetB.{id_col_name}"],
+        )
+
+        if self.isDefined(self.id_col):
+            knnjoin_df = knnjoin_df.select("datasetA", "datasetB", distCol)
+        else:
+            knnjoin_df = knnjoin_df.select(
+                knnjoin_df["datasetA"].dropFields(id_col_name).alias("datasetA"),
+                knnjoin_df["datasetB"].dropFields(id_col_name).alias("datasetB"),
+                distCol,
+            )
+
+        # restore relevant self variables
+        self._item_df_withid = estimator_item_df_withid
```

Comment:
This doesn't seem robust.  What if an error happens in some step before this restore?

Reply:
Yeah. This has been removed.